R/animint.R

```diff
@@ -557,34 +557,78 @@ saveLayer <- function(l, d, meta){
-            TRUE
-          }else if(all(counts == 1)){
-            FALSE #each chunk has only 1 row -- chunks are too small.
-          }else{
-            TRUE
+      names(can.chunk) <- subset.vec
+      ## Guess how big the chunk files will be, and reduce the number of
+      ## chunks if there are any that are too small.
+      tmp <- tempfile()
+      some.lines <- rbind(head(g.data), tail(g.data))
+      write.table(some.lines, tmp,
+                  col.names=FALSE,
+                  quote=FALSE, row.names=FALSE, sep="\t")
+      bytes <- file.size(tmp)
+      bytes.per.line <- bytes/nrow(some.lines)
+      bad.chunk <- function(){
+        if(all(!can.chunk))return(NULL)
+        can.chunk.cols <- subset.vec[can.chunk]
+        maybe.factors <- g.data[, can.chunk.cols, drop=FALSE]
+        for(N in names(maybe.factors)){
+          maybe.factors[[N]] <- paste(maybe.factors[[N]])
+        }
+        rows.per.chunk <- table(maybe.factors)
+        bytes.per.chunk <- rows.per.chunk * bytes.per.line
+        if(all(4096 < bytes.per.chunk))return(NULL)
+        ## If all of the tsv chunk files are greater than 4KB, then we
+        ## return NULL here to indicate that the current chunk
+        ## variables (indicated in can.chunk) are fine.
+
+        ## In other words, the compiler will not break a geom into
+        ## chunks if any of the resulting chunk tsv files is estimated
+        ## to be less than 4KB (of course, if the layer has very few
+        ## data overall, the compiler creates 1 file which may be less
+        ## than 4KB, but that is fine).
+        dim.byte.list <- list()
+        if(length(can.chunk.cols) == 1){
+          dim.byte.list[[can.chunk.cols]] <- sum(bytes.per.chunk)
+        }else{
+          for(dim.i in seq_along(can.chunk.cols)){
+            dim.name <- can.chunk.cols[[dim.i]]
+            dim.byte.list[[dim.name]] <-
+              apply(bytes.per.chunk, -dim.i, sum)
           }
         }
-        if(several.chunks){
-          nest.cols <- subset.vec[-can.chunk.i]
-          chunk.cols <- chunk.var
+        selector.df <-
+          data.frame(chunks.for=length(rows.per.chunk),
+                     chunks.without=sapply(dim.byte.list, length),
+                     min.bytes=sapply(dim.byte.list, min))
+        ## chunks.for is the number of chunks you get if you split the
+        ## data set using just this column. If it is 1, then it is
+        ## fine to chunk on this variable (since we certainly won't
+        ## make more than 1 small tsv file) and in fact we want to
+        ## chunk on this variable, since then this layer's data won't
+        ## be downloaded at first if it is not needed.
+        not.one <- subset(selector.df, 1 < chunks.for)
+        if(nrow(not.one) == 0){
+          NULL
         }else{
-          nest.cols <- subset.vec
-          chunk.cols <- NULL
+          rownames(not.one)[[which.max(not.one$min.bytes)]]
         }
       }
```

Comment:
by default the compiler now checks for showSelected variables that would create only 1 tsv file after splitting, and definitely includes them in chunk_order.

This fixes the bug with downloading too many chunks when the viz is first rendered.